common/src/main/java/com/wynntils/core/notifications/MessageContainer.java

```diff
@@ -9,17 +9,25 @@
 
 public class MessageContainer {
     protected TextRenderTask message;
+    private int messageCount;
 
     public MessageContainer(String message) {
         this.message = new TextRenderTask(message, TextRenderSetting.DEFAULT);
+        this.messageCount = 1;
     }
 
     public MessageContainer(TextRenderTask message) {
         this.message = message;
+        this.messageCount = 1;
     }
 
     public TextRenderTask getRenderTask() {
-        return message;
+        String sendableMessage = message.getText() + duplicateMessageBuilder(this);
+        return new TextRenderTask(sendableMessage, this.message.getSetting());
```

Comment:
I just noticed this! This is a no-no. This code is called every frame render. That means, we allocate a lot. Usually code like this looks just fine, but if you actually think about this, this is a perfectly hidden memory sink.

To fix this, keep a cached TextRenderTask with the duplicate message part, `renderedMessage`. Then, you can update it in `incrementMessageCount`, which is way less allocations.

Reply:
alright, i think i did what you wanted. my current implementation requires a nullcheck since i'm not able to figure out a better way to do it, so i would greatly appreciate any further pointers

Reply:
@san7890 You did what I asked of you. I did not think of this problem. :-( I apologize.